library/std/src/sys/unix/locks/queue_rwlock.rs

```diff
@@ -0,0 +1,492 @@
+                count = 0;
+            }
+        }
+    }
+
+    #[inline]
+    pub unsafe fn read_unlock(&self) {
+        match self.state.fetch_update(Release, Acquire, |state| {
+            if state.addr() & QUEUED == 0 {
+                let count = state.addr() - (SINGLE | LOCKED);
+                Some(if count > 0 { invalid_mut(count | LOCKED) } else { UNLOCKED })
+            } else {
+                None
+            }
+        }) {
+            Ok(_) => {}
+            Err(state) => unsafe { self.unlock_contended(state, true) },
+        }
+    }
+
+    #[inline]
+    pub unsafe fn write_unlock(&self) {
+        match self.state.compare_exchange(invalid_mut(LOCKED), UNLOCKED, Release, Acquire) {
+            Ok(_) => {}
+            // Since other threads cannot acquire the lock, the state can only
+            // have changed because there are threads queued on the lock.
+            Err(state) => unsafe { self.unlock_contended(state, false) },
+        }
+    }
+
+    /// # Safety
+    /// The lock must be locked by the current thread and threads must be queued on it.
+    #[cold]
+    unsafe fn unlock_contended(&self, mut state: State, read: bool) {
+        // Find the last node in the linked queue.
+        let tail = unsafe { find_tail(to_node(state)) };
+        let not_last = unsafe { read && tail.as_ref().decrement_count() };
+        if not_last {
+            // There are other lock owners, leave waking up the next waiters to them.
+            return;
+        }
+
+        // At this point, the `next` field on `tail` will always be null
+        // (invariant 2).
+
+        let next_read = unsafe { tail.as_ref().read };
+        if next_read {
+            // The next waiter is a reader. Just wake all threads.
+            //
+            // SAFETY:
+            // `current` is the head of a valid queue, which no thread except the
+            // the current can observe.
+            unsafe {
+                let mut current = to_node(self.state.swap(UNLOCKED, AcqRel));
+                loop {
+                    let next = current.as_ref().next.get();
+                    Node::complete(current);
+                    match next {
+                        Some(next) => current = next,
+                        None => break,
+                    }
+                }
+            }
+        } else {
+            // The next waiter is a writer. Remove it from the queue and wake it.
+            let prev = match unsafe { tail.as_ref().prev.get() } {
+                // If the lock was read-locked, multiple threads have invoked
+                // `find_tail` above. Therefore, it is possible that one  of
+                // them observed a newer state than this thread did, meaning
+                // there is a set `tail` field in a node before `state`. To
+                // make sure that the queue is valid after the link update
+                // below, reload the state and relink the queue.
```

Comment:
Shouldn't the queue still be valid here even if new nodes were added? What would change is a new head without a tail. This thread would update the tail of the old head, then a future `find_tail()` would store the updated tail into the new head. Reloading here seems like it does the same process but potentially earlier.

Reply:
Unfortunately, the resulting queue is not correct. Take the following execution order:

| Thread A           | Thread B           | Thread C             |
|-------------------:|:------------------:|:---------------------|
| h1 = load(state)   |                    |                      |
| t = find_tail(h1)  |                    | enqueue(state, node) |
|                    | h2 = load(state)   |                      |
|                    | t = find_tail(h2)  |                      |
|                    | decrement_count(t) |                      |
| decrement_count(t) |                    |                      |
| h.tail = t.prev    |                    |                      |

After these operations, the queue will look like:
```text
h2 -- h1 -- ...
|     |
|     current tail
|
old tail (removed)
```
so `find_tail` will find a dangling tail pointer. To preserve the invariant that, from the first node with a set `tail` field onwards, `prev` is initialized, the state needs to be reloaded here.

Reply:
Shouldnt the result of both `find_tail()`s be the same as the dequeue (which may make the tail dangling) only happens after the last decrement count?

Reply:
The result is the same, but it is cached in two different places. If the state is not reloaded, only the second value will be updated, so the next `find_tail` call will read the invalid cache value.

Reply:
Ah, sorry. Makes sense.

Reply:
Thanks for raising it! Rethinking it to explain it made me more confident that it's actually correct. :smile: